src/main/java/pro/trevor/tankgame/rule/impl/shared/PlayerRules.java

```diff
@@ -39,6 +39,34 @@ public class PlayerRules
         new DiscreteIntegerRange("gold",new HashSet<>(List.of(3, 5, 8, 10)))
     );
 
+    public static final PlayerActionRule<Tank> BuyActionWithGold(int action_cost)
+    {
+        if (action_cost <= 0)
+            throw new Error("Illegal Action Cost of " + action_cost + " gold. Must be positive and non-zero.");
+
+        return new PlayerActionRule<Tank>(
+            ActionKeys.BUY_ACTION,
+            (s, t, n) -> {
+                int gold_spent = toType(n[0], Integer.class);
+                return !t.isDead() && (t.getGold() >= gold_spent) && (gold_spent >= action_cost) && (gold_spent % action_cost == 0);
+            }, 
+            (s, t, n) -> {
+                int attempted_gold_spent = toType(n[0], Integer.class);
+                int bought_actions = attempted_gold_spent / action_cost;
+                int actual_gold_spent = bought_actions * action_cost;
+                assert attempted_gold_spent == actual_gold_spent;
+                
+                t.setActions(t.getActions() + bought_actions);
+                t.setGold(t.getGold() - actual_gold_spent);
+            }, 
+            new DiscreteIntegerRange("gold", new HashSet<>(List.of(1 * action_cost,
```

Comment:
This seems like it could be a for loop 0 to max actions... and maybe max actions should be a parameter.

Reply:
I thought about this, but I'm not sure how to do this without building in bad design because the rules are supposed to be independent, and I can't ask the Tank what it's max AP is at the moment. Thoughts?

Reply:
I was thinking of passing in the max number of actions for a specific rule set instead of having a global max.  The best case (which is a little more work) would be to write a range type that uses the current tanks max actions - actions.  Though this would only work with generic entity.

Reply:
This is fine for this review.

Reply:
Why can't this just be a rule for the case of purchasing one action? Or purchasing [1,n] actions where n is a parameter?

Also, maybe the numbers we display to the user should be the number of actions they are buying. The gold being spent does seem useful but I think the quantity of actions would be better to show. Ideally we would be able to show some meta-information i.e, both the quantity of actions and gold cost, but that is an issue for a separate PR.

Reply:
The UI could be updated so that the text doesn't match the value so we could display "3 gold -> 1 action" and send 1 action as the log entry pretty easily.

Reply:
Deferred